qiskit/circuit/quantumcircuit.py

```diff
@@ -5085,7 +5088,7 @@ def _qasm2_define_custom_operation(operation, existing_gate_names, gates_to_defi
                 instruction.operation, existing_gate_names, gates_to_define
             )
             bits_qasm = ",".join(qubit_labels[q] for q in instruction.qubits)
-            statements.append(f"{new_operation.qasm()} {bits_qasm};")
+            statements.append(f"{_instruction_qasm2(new_operation)} {bits_qasm};")
         body_qasm = " ".join(statements)
```

Comment:
Regarding "The ordering of these calls doesn't look right...". I made a couple of commits to simplify the code that show that all I've done is exchange a method call for a plain function call. However, I think this makes it clear that the call to `_instruction_qasm2` could be moved inside ` _qasm2_define_custom_operation`.  This would the code a bit less complex and add a bit more encapsulation. I think this may be what you are suggesting.

Reply:
Moving the hack down to a lower level feels worse in a sense. But, it is simpler and temporarily mutating something that should not change is distasteful. I didn't realize that it is more correct in the sense of your example. It's useful to make this clear